mojap_metadata/metadata/metadata.py

```diff
@@ -0,0 +1,115 @@
+
+    def __get__(self, obj, type=None) -> object:
+        return getattr(obj, self.name)
+
+    def __set__(self, obj, value) -> None:
+        setattr(obj, self.name, value)
+        obj.validate()
+
+
+class Metadata:
+    name = MetadataProperty()
+    description = MetadataProperty()
+    file_format = MetadataProperty()
+    sensitive = MetadataProperty()
+    columns = MetadataProperty()
+    primary_key = MetadataProperty()
+    partitions = MetadataProperty()
+
+    def __init__(
+        self,
+        name: str = "",
+        description: str = "",
+        file_format: str = "",
+        sensitive: bool = False,
+        columns: list = [],
+        primary_key: set = [],
+        partitions: set = [],
+    ) -> None:
+        self._name = name
+        self._description = description
+        self._file_format = file_format
+        self._sensitive = sensitive
+        self._columns = columns
+        self._primary_key = primary_key
+        self._partitions = partitions
+
+        with open("mojap_metadata/metadata/specs/table_schema.json") as file:
+            self._schema = json.load(file)
+
+        self.validate()
+
+    def validate(self):
+        jsonschema.validate(
+            instance=self.to_dict(include_schema=False), schema=self._schema
+        )
+        self._validate_list_attribute(attribute="primary_key", columns=self.primary_key)
+        self._validate_list_attribute(attribute="partitions", columns=self.partitions)
+
+    def _validate_list_attribute(self, attribute: str, columns: list) -> None:
+        if not isinstance(columns, list):
+            raise TypeError(f"'{attribute}' must be of type 'list'")
+        if not all([isinstance(column, str) for column in columns]):
+            raise TypeError(f"'{attribute}' must be a list of strings")
+        if not set(columns).issubset(
+            {column["name"] for column in self.columns if self.columns}
+        ):
+            raise ValueError(f"'All elements of '{attribute}' must be in self.columns")
+        if len(columns) != len(set(columns)):
+            raise ValueError(f"'All elements of '{attribute}' must be unique")
+
+    def from_dict(self, d: dict) -> None:
+        self.name = d.get("name", "")
+        self.description = d.get("description", "")
+        self.file_format = d.get("file_format", "")
+        self.sensitive = d.get("sensitive", False)
+        self.columns = d.get("columns", [])
+        self.primary_key = d.get("primary_key", [])
+        self.partitions = d.get("parititions", [])
+        if diff := set(d).difference(
+            {
+                "name",
+                "description",
+                "file_format",
+                "sensitive",
+                "columns",
+                "primary_key",
+                "partitions",
+            }
+        ):
+            warn(
+                f"Some properties will be ignored: {', '.join(sorted(diff))}",
+                UserWarning,
+            )
```

Comment:
Ideally you should be able to pass a dict to the object and it store the entire contents of that dict. This would mean that reading in a config, altering some parameters and then writing would be "lossless" if you wanted to just add additional unsupported properties of you config. 

This would require you to change the init, getters and setters but not that drastically.

Reply:
covered by a couple of existing issues